lib/UR/Role.pm

```diff
@@ -0,0 +1,732 @@
+
+=head1 SYNOPSIS
+
+  package My::Role;
+  role My::Role {
+      has => [
+          role_property => { is => 'String' },
+          another_prop  => { is => 'Integer' },
+      },
+      requires => ['class_method'],
+      excludes => ['Bad::Role'],
+  };
+  sub role_method { ... }
+
+
+  package My::Class;
+  class My::Class {
+      has => [
+          class_property => { is => 'Integer ' },
+      ],
+      roles => ['My::Role'],
+  };
+  sub class_method { ... }
+
+  my $obj = My::Class->new();
+  $obj->does('My::Role');  # true
+
+=head1 DESCRIPTION
+
+Roles are used to encapsulate a piece of behavior to be used in other classes.
+They have properties and methods that get melded into any class that composes
+them.  A Role can require any composing class to implement a list of methods
+or properties.
+
+Roles are not classes.  They can not be instantiated or inherited from.  They
+are composed into a class by listing their names in the C<roles> attribute of
+a class definition.
+
+=head2 Defining a Role
+
+Roles are defined with the C<role> keyword.  Their definition looks very
+similar to a class definition as described in L<UR::Object::Type::Initializer>.
+In particular, Roles have a C<has> section to define properties, and accept
+many class-meta attributes such as 'id_generator', 'valid_signals', and 'doc'.
+
+Roles may implement operator overloading via the 'use overload' mechanism.
+
+Roles also have unique atributes to declare restrictions on their use.
+
+=over 4
+
+=item requires
+
+A listref of property and method names that must appear in any class composing
+the Role.  Properties and methods defined in other roles or parent classes
+can satisfy a requirement.
+
+=item excludes
+
+A listref of Role names that may not be composed together with this Role.
+This is useful to declare incompatibilities between roles.
+
+=back
+
+=head2 Composing a Role
+
+Compose one or more Roles into a class using the 'roles' attribute in a class
+definition.
+
+  class My::Class {
+      roles => ['My::Role', 'Other::Role'],
+      is => ['Parent::Class'],
+      has => ['prop_a','prop_b'],
+  };
+
+Properties and meta-attributes from the Roles get copied into the composing
+class.  Subroutines defined in the Roles' namespaces are imported into the
+class's namespace.  Operator overloads defined in the Roles are applied to
+the class.
+
+=head3 Property and meta-attribute conflicts
+
+An exception is thrown if multiple Roles are composed together that define
+the same property, even if the composing class defines the same property in
+an attempt to override them.
+
+=head3 Method conflicts
+
+An exception is thrown if multiple Roles are composed together that
+define the same subroutine name unless the composing class's namespace also
+has a subroutine with the same name.  The class is expected to implement
+whatever behavior is required, maybe by calling one role's method or the
+other, both methods, neither, or anything else.
```

Comment:
Explicitly declaring these overrides sounds like a good idea, but we'd need to think about the syntax.  Declaring subs happens at compile time, and declaring classes (and consuming roles) happens at run-time.  Since we traditionally put class definitions at the top of the file, the roles would be consumed before any run-time ```override```s execute further down the file.

At first glance, it would also have the awkwardness of 'subtest', needing ```=> sub``` between the name and the block, and ending in a semicolon.  Devel::Declare might be able to help here.

This seems like a good place to use a function attribute - to declare it's overriding a role's implementation.

I'd like to find a different word than "override".  Override implies it has something to do with parent classes.  Maybe "supersede" or "replace"?